subsys/storage/storage_area/Kconfig

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2024 Laczen
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig STORAGE_AREA
+	bool "Storage area subsystem"
+	help
+	  Storage area subsystem.
+
+if STORAGE_AREA
+
+config STORAGE_AREA_VERIFY
+	bool "Verification support"
+	default y if DEBUG
+	help
+	  Verify if the definition of the storage area is
+	  valid for the used backend.
+
+config STORAGE_AREA_DISK
+	bool "Storage area on disk"
+	select DISK_ACCESS
+	help
+	  Use storage area on disk.
+
+config STORAGE_AREA_EEPROM
+	bool "Storage area on eeprom"
+	select EEPROM
+	help
+	  Use storage area on eeprom.
+
+config STORAGE_AREA_FLASH
+	bool "Storage area on flash"
+	select FLASH
+	help
+	  Use storage area on flash.
+
+config STORAGE_AREA_RAM
+	bool "Storage area on ram"
+	help
+	  Use storage area on ram.
+
+config STORAGE_AREA_STORE
+	bool "Storage area record storage"
+	help
+	  Use record storage on a storage area.
+
+if STORAGE_AREA_STORE
+
+config STORAGE_AREA_STORE_SEMAPHORE
+	bool "Thread safety using semaphore"
+	default y if (MULTITHREADING && !STORAGE_AREA_STORE_READONLY)
+	help
+	  Use a semaphore for multithreading.
+
+config STORAGE_AREA_STORE_DISABLECOMPACT
+	bool "Disable compacting"
+	help
+	  Disable compacting a storage area store, using the
+	  store as just a buffer. This will set the spare
+	  sector count to 0. This configuration is valid
+	  for ALL storage area stores.
+
+config STORAGE_AREA_STORE_READONLY
+	bool "Disable writing"
+	select STORAGE_AREA_STORE_DISABLECOMPACT
```

Comment:
Should this be an option per instance instead of a global option?

Reply:
The Kconfig options have been created to allow generating a minimal image size. These Kconfig options are associated with the library and not an instance,

In a typical application it is my expectation that there will be 3 different uses at the same time:
A. A read-only store containing some factory settings,
B. A circular buffer (without persistence requirement) type store that is used to store measurement data,
C. A store with persistence requirements e.g. to store settings,
These 3 uses are supported at the same time, and share code,

In a less typical application only the circular buffer might be used, code that is associated with the moving and persistence can be removed.

A different application might only want to use the store solution in a read-only way (e.g. a bootloader), in that case all code associated with writing can be removed.

Reply:
It might be better to put this into DT bindings. That way, it could be per-instance for anyone who has an unusual setup, but could easily be the normal setup too, with relatively little effort.

Reply:
There are no DT bindings, it should be passed to the macro definitions IMO.

Reply:
Regarding DT bindings, this is intentionally. IMHO this is a software layer and not a hardware layer and thus should not be configured by DT, but instead by the multi-instance configuration system that is being worked on.

It is however possible to use the provided macro's with data coming from DT, so if it is wanted to create storage area's based on DT this addition could be added easily (it is just a FOREACH_INST_...) for some backends (e.g. flash, ram), for other backends (eeprom, disk) it would need some kind of partitioning definition that is not available in DT. It is not the intention of this PR to go into DT changes.

Making the READ_ONLY, DISABLE_COMPACT instance based would mean splitting up the `storage_area_store` structs as well as all read/write routines and would duplicate code. This would result in a smaller (but dedicated) `storage_area_store` struct, but a much larger code size.

Reply:
No objection from me to having a global Kconfig option, but at the same time, it should be possible to have per-instance flags.

I don't see anything wrong with changing the macro implementation based on Kconfig settings - e.g. `COND_CODE_1()`. Similarly, a static inline wrapper could just use an `#ifdef` to simplify things.

Reply:
It was not the intention to create a read-only storage solution (nvs, fcb or zms also don't provide this). But since there are two potential users that think it would be good that a storage area store instance has a "flag" read-only I will have a look at how to best integrate this.